0.4.0 (#176)
ashleygwilliams/cargo-generate: CHANGELOG.md

# Changelog

## 🍕 0.4.0

  - ### ✨ Features
  
    - **Add .gen.toml for configuring include / exclude of files to template - [xortive], [pull/174]**
        

>         Adds support for a .gen.toml file which allows for templates to configure which files should be processed, either using a whitelist method (include), or a blacklist method (exclude), with include being preferred over exclude if both are present. This mirrors the behavior used by cargo for the configuration options of the same name in Cargo.toml, and uses some of the same code.

Reviewer: example + link to docs

       
        [xortive]: https://github.com/xortive
        [pull/174]: https://github.com/ashleygwilliams/cargo-generate/pull/174
        
    - **Filenames can now contains template tag(s) - [k0pernicus], [pull/161]**
        

>         Filenames that contain template tags will now be renamed!

Reviewer: also would give an example here and link to docs

        
        [k0pernicus]: https://github.com/k0pernicus
        [pull/161]: https://github.com/ashleygwilliams/cargo-generate/pull/161
        
  - ### 🤕 Fixes
  

>     - **Respect default branch name of template - [RotationMatrix], [pull/166]**

Reviewer: we should include an example here (i know it's a fix, but still nice)

# Changelog

## 🍕 0.4.0

  - ### ✨ Features

    - **Add .gen.toml for configuring include / exclude of files to template - [xortive], [pull/174]**
        

>         Adds support for a [.gen.toml](https://github.com/ashleygwilliams/cargo-generate/blob/master/README.md#include--exclude) file which allows for templates to configure which files should be processed, either using a whitelist method (include), or a blacklist method (exclude), with include being preferred over exclude if both are present. This mirrors the [behavior used by cargo](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields-optional) for the configuration options of the same name in Cargo.toml, and uses some of the same code.

Author: put the link where it will be once it's on master

>         Adds support for a [.gen.toml](https://github.com/ashleygwilliams/cargo-generate/blob/master/README.md#include--exclude) file which allows for templates to configure which files should be processed, either using a whitelist method (include), or a blacklist method (exclude), with include being preferred over exclude if both are present. This mirrors the [behavior used by cargo](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields-optional) for the configuration options of the same name in Cargo.toml, and uses some of the same code.

Reviewer 2: ```suggestion
        Adds support for the [.gen.toml](https://github.com/ashleygwilliams/cargo-generate/blob/master/README.md#include--exclude) file, which allows templates to configure which files should be processed, either using a whitelist method (`include`), or a blacklist method (`exclude`). When both `include` and `exclude` are present, `include` will be preferred. This mirrors [similar `include`/`exclude` behavior in `cargo`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields-optional), and uses some of the same code.
```

>         ```toml

Reviewer: let's show a valid example - this example is invalid which is confusing! so i'd add one that is include, one that is exclude, and then show the invalid one and say how it fails

        [template]
        include = ["Cargo.toml"]
        # include and exclude are exclusive, if both appear we will use include
        exclude = ["*.c"]
        ```
       
        [xortive]: https://github.com/xortive
        [pull/174]: https://github.com/ashleygwilliams/cargo-generate/pull/174

>     - **Filenames can now contains template tag(s) - [k0pernicus], [pull/161]**

Reviewer: this is not in the release anymore! make sure it's not in the README!

Author: Revert PR got rid of it

>         Filenames that contain template tags will now be renamed! For example, for a project named `awesome`, the filename `{{project_name}}.rs` will be transformed to `awesome.rs` during generation. <BEFORE MERGE: ADD DETAILS ABOUT HOW THIS INTERACTS WITH pull/174>

Author: need to add more details once 174 is merged

        
        [k0pernicus]: https://github.com/k0pernicus
        [pull/161]: https://github.com/ashleygwilliams/cargo-generate/pull/161

  - ### 🤕 Fixes

    - **Respect default branch name of template - [RotationMatrix], [pull/166]**
    

>         The `--branch` flag will now work as intended and set the initial HEAD to the specified branch name. For example, running `cargo generate --git <gh pages template> --branch gh-pages` will result in your generated project having it's default branch set to `gh-pages`.

Reviewer 2: ```suggestion
        The `--branch` flag will now work as intended and set the initial HEAD to the specified branch name. For example, running `cargo generate --git <gh pages template> --branch gh-pages` will set your generated project's default branch to `gh-pages`.
```

Reviewer 2: What does "default branch" mean in this context? Is it actually changing from Git's default behavior of using the branch `master` as default, or just setting the initial branch for when you `cd` into the project?

Author: "default branch" is a bit of a misnomer - Only GitHub has the idea of one. What this is actually doing is setting the initial branch for when you CD into the project.

        [cyclowns]: https://github.com/cyclowns
        [pull/167]: https://github.com/ashleygwilliams/cargo-generate/pull/167

> - ### 🛠️ Maintenance

Reviewer: nice catch!
